lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -48,35 +49,43 @@ def __init__(self, config: PbenchServerConfig, logger: logging.Logger):
             role=API_OPERATION.READ,
         )
 
-    def get_paginated_obj(self, query: Query, json: JSON, url: str):
+    def get_paginated_obj(
+        self, query: Query, json: JSON, url: str
+    ) -> Tuple[List, Dict[str, str]]:
         """
-        Helper function to return a paginated object containing slice of dataset
-        (constructed according to the page limit and a page number), next page url,
-        and total items count.
-        TODO: There are multiple ways to optimize the functionality
-            1. Use of caching mechanism to temporarily store the queried datasets
-            instead of querying the database with every page request.
-            2. Use of unique pointers to store the last returned row and then use
-            this pointer in subsequent page request instead of initial start to
+        Helper function to return a slice of datasets (constructed according to the
+        user specified limit and an offset number) and a paginated object containing next page
+        url and total items count.
+        E.g. specifying the following limit and offset values will result in the corresponding
+        dataset slice:
+        "limit": 10, "offset": 20 -> dataset[20: 30]
+        "limit": 10 -> dataset[0: 10]
+        "offset": 20 -> dataset[20: total_items_count]
+        TODO: We may need to optimize the pagination
+            e.g Use of unique pointers to record the last returned row and then use
+            this pointer in subsequent page request instead of an initial start to
             narrow down the result.
         """
         paginated_result = {}
         total_count = query.count()
-        page_limit = json.get(
-            "page_limit", int(self.config.server_config.get("page_limit"))
-        )
-        page_no = json.get("page_no", 1)
-        items = (
-            query.order_by(Dataset.name)
-            .limit(page_limit)
-            .offset((page_no - 1) * page_limit)
-            .all()
-        )
-        json["page_no"] = page_no + 1
-        parsed_url = urlparse(url)
-        if len(items) < page_limit:
+        query = query.order_by(Dataset.name)
+
+        # Get the user specified limit, otherwise return all the items
+        limit = json.get("limit", total_count + 1)
+        query = query.limit(limit)
```

Comment:
This is OK, but wouldn't it be better to just skip the `limit()` call if the user specifies no limit?

Reply:
In effect, they are the same thing but yeah we could avoid using the limit, this approach just avoids the conditional statement that comes when avoiding the limit.

Reply:
Yeah, it depends on which is more expensive, the call to `query.limit()` or the conditional statement (where the "cost" is not only in terms of execution time but also code complexity, testing, etc.).

Also, if we don't make the call to `limit()` then we don't have to worry about whether the default should be `total_count` or `total_count + 1`.  🙃  (I'm _pretty_ sure that the `+1` is superfluous and a waste....)